// @databases/push-to-async-iterable/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

const queue_1 = require("@databases/queue");

function pushToAsyncIterable(getStream) {
  const queue = new queue_1.AsyncQueue();
  let bufferSize = 0;
  let paused = false;
  let ended = false;
  const stream = getStream({
    onData(value) {
      if (!ended) {
        queue.push({
          done: false,
          value
        });
        bufferSize++;

        if (!paused && bufferSize >= stream.highWaterMark) {
          paused = true;
          stream.pause();
        }
      }
    },

    onError(err) {
      if (!ended) {
        ended = true;
        queue.push({
          done: true,
          err
        });
      }
    },

    onEnd() {
      if (!ended) {
        ended = true;
        queue.push({
          done: true,
          err: undefined
        });
      }
    }

  });
  return {
    async next() {
      bufferSize--;

      if (paused && bufferSize < stream.highWaterMark) {
        paused = false;
        stream.resume();
      }

      const next = await queue.shift();

      if (next.done && next.err) {
        throw next.err;
      } else if (next.done) {
        return {
          done: true,
          value: undefined
        };
      } else {
        return next;
      }
    },

    async return() {
      stream.dispose();
      return {
        done: true,
        value: undefined
      };
    },

    async throw(e) {
      stream.dispose();
      throw e;
    },

    [Symbol.asyncIterator]() {
      // tslint:disable-next-line no-invalid-this
      return this;
    }

  };
}

exports.default = pushToAsyncIterable;
