// @databases/pg-config/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PgTypesPrimaryKeyTypeMode = exports.PgTypesEnumTypeMode = exports.PgTypesDomainTypeMode = exports.DEFAULT_CONFIG = exports.readPgConfigSync = exports.getPgConfigSync = exports.getPgConfig = void 0;
const cosmiconfig_1 = require("cosmiconfig");
const PgConfig_1 = require("./PgConfig");
const asyncExplorer = (0, cosmiconfig_1.cosmiconfig)('pg');
const syncExplorer = (0, cosmiconfig_1.cosmiconfigSync)('pg');
async function getPgConfig(searchFrom) {
    return parseResult(await asyncExplorer.search(searchFrom));
}
exports.getPgConfig = getPgConfig;
function getPgConfigSync(searchFrom) {
    return parseResult(syncExplorer.search(searchFrom));
}
exports.getPgConfigSync = getPgConfigSync;
function readPgConfigSync(filename) {
    return parseResult(syncExplorer.load(filename));
}
exports.readPgConfigSync = readPgConfigSync;
function parseResult(result) {
    return PgConfig_1.PgConfigSchema.parse(result ? result.config : {});
}
exports.DEFAULT_CONFIG = PgConfig_1.PgConfigSchema.parse({});
var PgConfig_2 = require("./PgConfig");
Object.defineProperty(exports, "PgTypesDomainTypeMode", { enumerable: true, get: function () { return PgConfig_2.PgTypesDomainTypeMode; } });
Object.defineProperty(exports, "PgTypesEnumTypeMode", { enumerable: true, get: function () { return PgConfig_2.PgTypesEnumTypeMode; } });
Object.defineProperty(exports, "PgTypesPrimaryKeyTypeMode", { enumerable: true, get: function () { return PgConfig_2.PgTypesPrimaryKeyTypeMode; } });
