// @databases/connection-pool/lib/IPoolRecord.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPoolRecord = exports.isDisposedPoolRecord = exports.isIdlePoolRecord = exports.isActivePoolRecord = exports.setRecordState = exports.PoolRecordState = void 0;
var PoolRecordState;
(function (PoolRecordState) {
    PoolRecordState[PoolRecordState["Idle"] = 0] = "Idle";
    PoolRecordState[PoolRecordState["Active"] = 1] = "Active";
    PoolRecordState[PoolRecordState["Disposed"] = 2] = "Disposed";
})(PoolRecordState = exports.PoolRecordState || (exports.PoolRecordState = {}));
function setRecordState(record, state, onTimeout, timeoutMilliseconds) {
    if (record.idleTimeout !== undefined) {
        clearTimeout(record.idleTimeout);
    }
    if (onTimeout && timeoutMilliseconds !== Infinity) {
        record.idleTimeout = setTimeout(onTimeout, timeoutMilliseconds, record);
    }
    record.state = state;
    switch (state) {
        case PoolRecordState.Active:
            record.activateCount++;
            break;
        case PoolRecordState.Disposed:
            record.connection = undefined;
            break;
    }
    return record;
}
exports.setRecordState = setRecordState;
function isActivePoolRecord(record) {
    return record.state === PoolRecordState.Active;
}
exports.isActivePoolRecord = isActivePoolRecord;
function isIdlePoolRecord(record) {
    return record.state === PoolRecordState.Idle;
}
exports.isIdlePoolRecord = isIdlePoolRecord;
function isDisposedPoolRecord(record) {
    return record.state === PoolRecordState.Disposed;
}
exports.isDisposedPoolRecord = isDisposedPoolRecord;
class PoolRecord {
    constructor(connection) {
        this.state = PoolRecordState.Idle;
        this.activateCount = 0;
        this.shouldDestroy = false;
        this.connection = connection;
    }
}
function getPoolRecord(connection) {
    const record = new PoolRecord(connection);
    if (!isIdlePoolRecord(record)) {
        throw new Error('Newly created records should alays be idle');
    }
    return record;
}
exports.getPoolRecord = getPoolRecord;
