// @databases/pg-errors/lib/SQLErrorCode.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var SQLErrorCode;

(function (SQLErrorCode) {
  SQLErrorCode["SUCCESSFUL_COMPLETION"] = "00000";
  SQLErrorCode["WARNING"] = "01000";
  SQLErrorCode["DYNAMIC_RESULT_SETS_RETURNED"] = "0100C";
  SQLErrorCode["IMPLICIT_ZERO_BIT_PADDING"] = "01008";
  SQLErrorCode["NULL_VALUE_ELIMINATED_IN_SET_FUNCTION"] = "01003";
  SQLErrorCode["PRIVILEGE_NOT_GRANTED"] = "01007";
  SQLErrorCode["PRIVILEGE_NOT_REVOKED"] = "01006";
  SQLErrorCode["STRING_DATA_RIGHT_TRUNCATION"] = "22001";
  SQLErrorCode["DEPRECATED_FEATURE"] = "01P01";
  SQLErrorCode["NO_DATA"] = "02000";
  SQLErrorCode["NO_ADDITIONAL_DYNAMIC_RESULT_SETS_RETURNED"] = "02001";
  SQLErrorCode["SQL_STATEMENT_NOT_YET_COMPLETE"] = "03000";
  SQLErrorCode["CONNECTION_EXCEPTION"] = "08000";
  SQLErrorCode["CONNECTION_DOES_NOT_EXIST"] = "08003";
  SQLErrorCode["CONNECTION_FAILURE"] = "08006";
  SQLErrorCode["SQLCLIENT_UNABLE_TO_ESTABLISH_SQLCONNECTION"] = "08001";
  SQLErrorCode["SQLSERVER_REJECTED_ESTABLISHMENT_OF_SQLCONNECTION"] = "08004";
  SQLErrorCode["TRANSACTION_RESOLUTION_UNKNOWN"] = "08007";
  SQLErrorCode["PROTOCOL_VIOLATION"] = "08P01";
  SQLErrorCode["TRIGGERED_ACTION_EXCEPTION"] = "09000";
  SQLErrorCode["FEATURE_NOT_SUPPORTED"] = "0A000";
  SQLErrorCode["INVALID_TRANSACTION_INITIATION"] = "0B000";
  SQLErrorCode["LOCATOR_EXCEPTION"] = "0F000";
  SQLErrorCode["INVALID_LOCATOR_SPECIFICATION"] = "0F001";
  SQLErrorCode["INVALID_GRANTOR"] = "0L000";
  SQLErrorCode["INVALID_GRANT_OPERATION"] = "0LP01";
  SQLErrorCode["INVALID_ROLE_SPECIFICATION"] = "0P000";
  SQLErrorCode["DIAGNOSTICS_EXCEPTION"] = "0Z000";
  SQLErrorCode["STACKED_DIAGNOSTICS_ACCESSED_WITHOUT_ACTIVE_HANDLER"] = "0Z002";
  SQLErrorCode["CASE_NOT_FOUND"] = "20000";
  SQLErrorCode["CARDINALITY_VIOLATION"] = "21000";
  SQLErrorCode["DATA_EXCEPTION"] = "22000";
  SQLErrorCode["ARRAY_SUBSCRIPT_ERROR"] = "2202E";
  SQLErrorCode["CHARACTER_NOT_IN_REPERTOIRE"] = "22021";
  SQLErrorCode["DATETIME_FIELD_OVERFLOW"] = "22008";
  SQLErrorCode["DIVISION_BY_ZERO"] = "22012";
  SQLErrorCode["ERROR_IN_ASSIGNMENT"] = "22005";
  SQLErrorCode["ESCAPE_CHARACTER_CONFLICT"] = "2200B";
  SQLErrorCode["INDICATOR_OVERFLOW"] = "22022";
  SQLErrorCode["INTERVAL_FIELD_OVERFLOW"] = "22015";
  SQLErrorCode["INVALID_ARGUMENT_FOR_LOGARITHM"] = "2201E";
  SQLErrorCode["INVALID_ARGUMENT_FOR_NTILE_FUNCTION"] = "22014";
  SQLErrorCode["INVALID_ARGUMENT_FOR_NTH_VALUE_FUNCTION"] = "22016";
  SQLErrorCode["INVALID_ARGUMENT_FOR_POWER_FUNCTION"] = "2201F";
  SQLErrorCode["INVALID_ARGUMENT_FOR_WIDTH_BUCKET_FUNCTION"] = "2201G";
  SQLErrorCode["INVALID_CHARACTER_VALUE_FOR_CAST"] = "22018";
  SQLErrorCode["INVALID_DATETIME_FORMAT"] = "22007";
  SQLErrorCode["INVALID_ESCAPE_CHARACTER"] = "22019";
  SQLErrorCode["INVALID_ESCAPE_OCTET"] = "2200D";
  SQLErrorCode["INVALID_ESCAPE_SEQUENCE"] = "22025";
  SQLErrorCode["NONSTANDARD_USE_OF_ESCAPE_CHARACTER"] = "22P06";
  SQLErrorCode["INVALID_INDICATOR_PARAMETER_VALUE"] = "22010";
  SQLErrorCode["INVALID_PARAMETER_VALUE"] = "22023";
  SQLErrorCode["INVALID_REGULAR_EXPRESSION"] = "2201B";
  SQLErrorCode["INVALID_ROW_COUNT_IN_LIMIT_CLAUSE"] = "2201W";
  SQLErrorCode["INVALID_ROW_COUNT_IN_RESULT_OFFSET_CLAUSE"] = "2201X";
  SQLErrorCode["INVALID_TIME_ZONE_DISPLACEMENT_VALUE"] = "22009";
  SQLErrorCode["INVALID_USE_OF_ESCAPE_CHARACTER"] = "2200C";
  SQLErrorCode["MOST_SPECIFIC_TYPE_MISMATCH"] = "2200G";
  SQLErrorCode["NULL_VALUE_NOT_ALLOWED"] = "39004";
  SQLErrorCode["NULL_VALUE_NO_INDICATOR_PARAMETER"] = "22002";
  SQLErrorCode["NUMERIC_VALUE_OUT_OF_RANGE"] = "22003";
  SQLErrorCode["STRING_DATA_LENGTH_MISMATCH"] = "22026";
  SQLErrorCode["SUBSTRING_ERROR"] = "22011";
  SQLErrorCode["TRIM_ERROR"] = "22027";
  SQLErrorCode["UNTERMINATED_C_STRING"] = "22024";
  SQLErrorCode["ZERO_LENGTH_CHARACTER_STRING"] = "2200F";
  SQLErrorCode["FLOATING_POINT_EXCEPTION"] = "22P01";
  SQLErrorCode["INVALID_TEXT_REPRESENTATION"] = "22P02";
  SQLErrorCode["INVALID_BINARY_REPRESENTATION"] = "22P03";
  SQLErrorCode["BAD_COPY_FILE_FORMAT"] = "22P04";
  SQLErrorCode["UNTRANSLATABLE_CHARACTER"] = "22P05";
  SQLErrorCode["NOT_AN_XML_DOCUMENT"] = "2200L";
  SQLErrorCode["INVALID_XML_DOCUMENT"] = "2200M";
  SQLErrorCode["INVALID_XML_CONTENT"] = "2200N";
  SQLErrorCode["INVALID_XML_COMMENT"] = "2200S";
  SQLErrorCode["INVALID_XML_PROCESSING_INSTRUCTION"] = "2200T";
  SQLErrorCode["INTEGRITY_CONSTRAINT_VIOLATION"] = "23000";
  SQLErrorCode["RESTRICT_VIOLATION"] = "23001";
  SQLErrorCode["NOT_NULL_VIOLATION"] = "23502";
  SQLErrorCode["FOREIGN_KEY_VIOLATION"] = "23503";
  SQLErrorCode["UNIQUE_VIOLATION"] = "23505";
  SQLErrorCode["CHECK_VIOLATION"] = "23514";
  SQLErrorCode["EXCLUSION_VIOLATION"] = "23P01";
  SQLErrorCode["INVALID_CURSOR_STATE"] = "24000";
  SQLErrorCode["INVALID_TRANSACTION_STATE"] = "25000";
  SQLErrorCode["ACTIVE_SQL_TRANSACTION"] = "25001";
  SQLErrorCode["BRANCH_TRANSACTION_ALREADY_ACTIVE"] = "25002";
  SQLErrorCode["HELD_CURSOR_REQUIRES_SAME_ISOLATION_LEVEL"] = "25008";
  SQLErrorCode["INAPPROPRIATE_ACCESS_MODE_FOR_BRANCH_TRANSACTION"] = "25003";
  SQLErrorCode["INAPPROPRIATE_ISOLATION_LEVEL_FOR_BRANCH_TRANSACTION"] = "25004";
  SQLErrorCode["NO_ACTIVE_SQL_TRANSACTION_FOR_BRANCH_TRANSACTION"] = "25005";
  SQLErrorCode["READ_ONLY_SQL_TRANSACTION"] = "25006";
  SQLErrorCode["SCHEMA_AND_DATA_STATEMENT_MIXING_NOT_SUPPORTED"] = "25007";
  SQLErrorCode["NO_ACTIVE_SQL_TRANSACTION"] = "25P01";
  SQLErrorCode["IN_FAILED_SQL_TRANSACTION"] = "25P02";
  SQLErrorCode["INVALID_SQL_STATEMENT_NAME"] = "26000";
  SQLErrorCode["TRIGGERED_DATA_CHANGE_VIOLATION"] = "27000";
  SQLErrorCode["INVALID_AUTHORIZATION_SPECIFICATION"] = "28000";
  SQLErrorCode["INVALID_PASSWORD"] = "28P01";
  SQLErrorCode["DEPENDENT_PRIVILEGE_DESCRIPTORS_STILL_EXIST"] = "2B000";
  SQLErrorCode["DEPENDENT_OBJECTS_STILL_EXIST"] = "2BP01";
  SQLErrorCode["INVALID_TRANSACTION_TERMINATION"] = "2D000";
  SQLErrorCode["SQL_ROUTINE_EXCEPTION"] = "2F000";
  SQLErrorCode["FUNCTION_EXECUTED_NO_RETURN_STATEMENT"] = "2F005";
  SQLErrorCode["MODIFYING_SQL_DATA_NOT_PERMITTED"] = "38002";
  SQLErrorCode["PROHIBITED_SQL_STATEMENT_ATTEMPTED"] = "38003";
  SQLErrorCode["READING_SQL_DATA_NOT_PERMITTED"] = "38004";
  SQLErrorCode["INVALID_CURSOR_NAME"] = "34000";
  SQLErrorCode["EXTERNAL_ROUTINE_EXCEPTION"] = "38000";
  SQLErrorCode["CONTAINING_SQL_NOT_PERMITTED"] = "38001";
  SQLErrorCode["EXTERNAL_ROUTINE_INVOCATION_EXCEPTION"] = "39000";
  SQLErrorCode["INVALID_SQLSTATE_RETURNED"] = "39001";
  SQLErrorCode["TRIGGER_PROTOCOL_VIOLATED"] = "39P01";
  SQLErrorCode["SRF_PROTOCOL_VIOLATED"] = "39P02";
  SQLErrorCode["SAVEPOINT_EXCEPTION"] = "3B000";
  SQLErrorCode["INVALID_SAVEPOINT_SPECIFICATION"] = "3B001";
  SQLErrorCode["INVALID_CATALOG_NAME"] = "3D000";
  SQLErrorCode["INVALID_SCHEMA_NAME"] = "3F000";
  SQLErrorCode["TRANSACTION_ROLLBACK"] = "40000";
  SQLErrorCode["TRANSACTION_INTEGRITY_CONSTRAINT_VIOLATION"] = "40002";
  SQLErrorCode["SERIALIZATION_FAILURE"] = "40001";
  SQLErrorCode["STATEMENT_COMPLETION_UNKNOWN"] = "40003";
  SQLErrorCode["DEADLOCK_DETECTED"] = "40P01";
  SQLErrorCode["SYNTAX_ERROR_OR_ACCESS_RULE_VIOLATION"] = "42000";
  SQLErrorCode["SYNTAX_ERROR"] = "42601";
  SQLErrorCode["INSUFFICIENT_PRIVILEGE"] = "42501";
  SQLErrorCode["CANNOT_COERCE"] = "42846";
  SQLErrorCode["GROUPING_ERROR"] = "42803";
  SQLErrorCode["WINDOWING_ERROR"] = "42P20";
  SQLErrorCode["INVALID_RECURSION"] = "42P19";
  SQLErrorCode["INVALID_FOREIGN_KEY"] = "42830";
  SQLErrorCode["INVALID_NAME"] = "42602";
  SQLErrorCode["NAME_TOO_LONG"] = "42622";
  SQLErrorCode["RESERVED_NAME"] = "42939";
  SQLErrorCode["DATATYPE_MISMATCH"] = "42804";
  SQLErrorCode["INDETERMINATE_DATATYPE"] = "42P18";
  SQLErrorCode["COLLATION_MISMATCH"] = "42P21";
  SQLErrorCode["INDETERMINATE_COLLATION"] = "42P22";
  SQLErrorCode["WRONG_OBJECT_TYPE"] = "42809";
  SQLErrorCode["UNDEFINED_COLUMN"] = "42703";
  SQLErrorCode["UNDEFINED_FUNCTION"] = "42883";
  SQLErrorCode["UNDEFINED_TABLE"] = "42P01";
  SQLErrorCode["UNDEFINED_PARAMETER"] = "42P02";
  SQLErrorCode["UNDEFINED_OBJECT"] = "42704";
  SQLErrorCode["DUPLICATE_COLUMN"] = "42701";
  SQLErrorCode["DUPLICATE_CURSOR"] = "42P03";
  SQLErrorCode["DUPLICATE_DATABASE"] = "42P04";
  SQLErrorCode["DUPLICATE_FUNCTION"] = "42723";
  SQLErrorCode["DUPLICATE_PREPARED_STATEMENT"] = "42P05";
  SQLErrorCode["DUPLICATE_SCHEMA"] = "42P06";
  SQLErrorCode["DUPLICATE_TABLE"] = "42P07";
  SQLErrorCode["DUPLICATE_ALIAS"] = "42712";
  SQLErrorCode["DUPLICATE_OBJECT"] = "42710";
  SQLErrorCode["AMBIGUOUS_COLUMN"] = "42702";
  SQLErrorCode["AMBIGUOUS_FUNCTION"] = "42725";
  SQLErrorCode["AMBIGUOUS_PARAMETER"] = "42P08";
  SQLErrorCode["AMBIGUOUS_ALIAS"] = "42P09";
  SQLErrorCode["INVALID_COLUMN_REFERENCE"] = "42P10";
  SQLErrorCode["INVALID_COLUMN_DEFINITION"] = "42611";
  SQLErrorCode["INVALID_CURSOR_DEFINITION"] = "42P11";
  SQLErrorCode["INVALID_DATABASE_DEFINITION"] = "42P12";
  SQLErrorCode["INVALID_FUNCTION_DEFINITION"] = "42P13";
  SQLErrorCode["INVALID_PREPARED_STATEMENT_DEFINITION"] = "42P14";
  SQLErrorCode["INVALID_SCHEMA_DEFINITION"] = "42P15";
  SQLErrorCode["INVALID_TABLE_DEFINITION"] = "42P16";
  SQLErrorCode["INVALID_OBJECT_DEFINITION"] = "42P17";
  SQLErrorCode["WITH_CHECK_OPTION_VIOLATION"] = "44000";
  SQLErrorCode["INSUFFICIENT_RESOURCES"] = "53000";
  SQLErrorCode["DISK_FULL"] = "53100";
  SQLErrorCode["OUT_OF_MEMORY"] = "53200";
  SQLErrorCode["TOO_MANY_CONNECTIONS"] = "53300";
  SQLErrorCode["CONFIGURATION_LIMIT_EXCEEDED"] = "53400";
  SQLErrorCode["PROGRAM_LIMIT_EXCEEDED"] = "54000";
  SQLErrorCode["STATEMENT_TOO_COMPLEX"] = "54001";
  SQLErrorCode["TOO_MANY_COLUMNS"] = "54011";
  SQLErrorCode["TOO_MANY_ARGUMENTS"] = "54023";
  SQLErrorCode["OBJECT_NOT_IN_PREREQUISITE_STATE"] = "55000";
  SQLErrorCode["OBJECT_IN_USE"] = "55006";
  SQLErrorCode["CANT_CHANGE_RUNTIME_PARAM"] = "55P02";
  SQLErrorCode["LOCK_NOT_AVAILABLE"] = "55P03";
  SQLErrorCode["OPERATOR_INTERVENTION"] = "57000";
  SQLErrorCode["QUERY_CANCELED"] = "57014";
  SQLErrorCode["ADMIN_SHUTDOWN"] = "57P01";
  SQLErrorCode["CRASH_SHUTDOWN"] = "57P02";
  SQLErrorCode["CANNOT_CONNECT_NOW"] = "57P03";
  SQLErrorCode["DATABASE_DROPPED"] = "57P04";
  SQLErrorCode["SYSTEM_ERROR"] = "58000";
  SQLErrorCode["IO_ERROR"] = "58030";
  SQLErrorCode["UNDEFINED_FILE"] = "58P01";
  SQLErrorCode["DUPLICATE_FILE"] = "58P02";
  SQLErrorCode["CONFIG_FILE_ERROR"] = "F0000";
  SQLErrorCode["LOCK_FILE_EXISTS"] = "F0001";
  SQLErrorCode["FDW_ERROR"] = "HV000";
  SQLErrorCode["FDW_COLUMN_NAME_NOT_FOUND"] = "HV005";
  SQLErrorCode["FDW_DYNAMIC_PARAMETER_VALUE_NEEDED"] = "HV002";
  SQLErrorCode["FDW_FUNCTION_SEQUENCE_ERROR"] = "HV010";
  SQLErrorCode["FDW_INCONSISTENT_DESCRIPTOR_INFORMATION"] = "HV021";
  SQLErrorCode["FDW_INVALID_ATTRIBUTE_VALUE"] = "HV024";
  SQLErrorCode["FDW_INVALID_COLUMN_NAME"] = "HV007";
  SQLErrorCode["FDW_INVALID_COLUMN_NUMBER"] = "HV008";
  SQLErrorCode["FDW_INVALID_DATA_TYPE"] = "HV004";
  SQLErrorCode["FDW_INVALID_DATA_TYPE_DESCRIPTORS"] = "HV006";
  SQLErrorCode["FDW_INVALID_DESCRIPTOR_FIELD_IDENTIFIER"] = "HV091";
  SQLErrorCode["FDW_INVALID_HANDLE"] = "HV00B";
  SQLErrorCode["FDW_INVALID_OPTION_INDEX"] = "HV00C";
  SQLErrorCode["FDW_INVALID_OPTION_NAME"] = "HV00D";
  SQLErrorCode["FDW_INVALID_STRING_LENGTH_OR_BUFFER_LENGTH"] = "HV090";
  SQLErrorCode["FDW_INVALID_STRING_FORMAT"] = "HV00A";
  SQLErrorCode["FDW_INVALID_USE_OF_NULL_POINTER"] = "HV009";
  SQLErrorCode["FDW_TOO_MANY_HANDLES"] = "HV014";
  SQLErrorCode["FDW_OUT_OF_MEMORY"] = "HV001";
  SQLErrorCode["FDW_NO_SCHEMAS"] = "HV00P";
  SQLErrorCode["FDW_OPTION_NAME_NOT_FOUND"] = "HV00J";
  SQLErrorCode["FDW_REPLY_HANDLE"] = "HV00K";
  SQLErrorCode["FDW_SCHEMA_NOT_FOUND"] = "HV00Q";
  SQLErrorCode["FDW_TABLE_NOT_FOUND"] = "HV00R";
  SQLErrorCode["FDW_UNABLE_TO_CREATE_EXECUTION"] = "HV00L";
  SQLErrorCode["FDW_UNABLE_TO_CREATE_REPLY"] = "HV00M";
  SQLErrorCode["FDW_UNABLE_TO_ESTABLISH_CONNECTION"] = "HV00N";
  SQLErrorCode["PLPGSQL_ERROR"] = "P0000";
  SQLErrorCode["RAISE_EXCEPTION"] = "P0001";
  SQLErrorCode["NO_DATA_FOUND"] = "P0002";
  SQLErrorCode["TOO_MANY_ROWS"] = "P0003";
  SQLErrorCode["INTERNAL_ERROR"] = "XX000";
  SQLErrorCode["DATA_CORRUPTED"] = "XX001";
  SQLErrorCode["INDEX_CORRUPTED"] = "XX002";
})(SQLErrorCode = exports.SQLErrorCode || (exports.SQLErrorCode = {}));

exports.default = SQLErrorCode;
exports.SQLErrorCodes = new Set([SQLErrorCode.SUCCESSFUL_COMPLETION, SQLErrorCode.WARNING, SQLErrorCode.DYNAMIC_RESULT_SETS_RETURNED, SQLErrorCode.IMPLICIT_ZERO_BIT_PADDING, SQLErrorCode.NULL_VALUE_ELIMINATED_IN_SET_FUNCTION, SQLErrorCode.PRIVILEGE_NOT_GRANTED, SQLErrorCode.PRIVILEGE_NOT_REVOKED, SQLErrorCode.STRING_DATA_RIGHT_TRUNCATION, SQLErrorCode.DEPRECATED_FEATURE, SQLErrorCode.NO_DATA, SQLErrorCode.NO_ADDITIONAL_DYNAMIC_RESULT_SETS_RETURNED, SQLErrorCode.SQL_STATEMENT_NOT_YET_COMPLETE, SQLErrorCode.CONNECTION_EXCEPTION, SQLErrorCode.CONNECTION_DOES_NOT_EXIST, SQLErrorCode.CONNECTION_FAILURE, SQLErrorCode.SQLCLIENT_UNABLE_TO_ESTABLISH_SQLCONNECTION, SQLErrorCode.SQLSERVER_REJECTED_ESTABLISHMENT_OF_SQLCONNECTION, SQLErrorCode.TRANSACTION_RESOLUTION_UNKNOWN, SQLErrorCode.PROTOCOL_VIOLATION, SQLErrorCode.TRIGGERED_ACTION_EXCEPTION, SQLErrorCode.FEATURE_NOT_SUPPORTED, SQLErrorCode.INVALID_TRANSACTION_INITIATION, SQLErrorCode.LOCATOR_EXCEPTION, SQLErrorCode.INVALID_LOCATOR_SPECIFICATION, SQLErrorCode.INVALID_GRANTOR, SQLErrorCode.INVALID_GRANT_OPERATION, SQLErrorCode.INVALID_ROLE_SPECIFICATION, SQLErrorCode.DIAGNOSTICS_EXCEPTION, SQLErrorCode.STACKED_DIAGNOSTICS_ACCESSED_WITHOUT_ACTIVE_HANDLER, SQLErrorCode.CASE_NOT_FOUND, SQLErrorCode.CARDINALITY_VIOLATION, SQLErrorCode.DATA_EXCEPTION, SQLErrorCode.ARRAY_SUBSCRIPT_ERROR, SQLErrorCode.CHARACTER_NOT_IN_REPERTOIRE, SQLErrorCode.DATETIME_FIELD_OVERFLOW, SQLErrorCode.DIVISION_BY_ZERO, SQLErrorCode.ERROR_IN_ASSIGNMENT, SQLErrorCode.ESCAPE_CHARACTER_CONFLICT, SQLErrorCode.INDICATOR_OVERFLOW, SQLErrorCode.INTERVAL_FIELD_OVERFLOW, SQLErrorCode.INVALID_ARGUMENT_FOR_LOGARITHM, SQLErrorCode.INVALID_ARGUMENT_FOR_NTILE_FUNCTION, SQLErrorCode.INVALID_ARGUMENT_FOR_NTH_VALUE_FUNCTION, SQLErrorCode.INVALID_ARGUMENT_FOR_POWER_FUNCTION, SQLErrorCode.INVALID_ARGUMENT_FOR_WIDTH_BUCKET_FUNCTION, SQLErrorCode.INVALID_CHARACTER_VALUE_FOR_CAST, SQLErrorCode.INVALID_DATETIME_FORMAT, SQLErrorCode.INVALID_ESCAPE_CHARACTER, SQLErrorCode.INVALID_ESCAPE_OCTET, SQLErrorCode.INVALID_ESCAPE_SEQUENCE, SQLErrorCode.NONSTANDARD_USE_OF_ESCAPE_CHARACTER, SQLErrorCode.INVALID_INDICATOR_PARAMETER_VALUE, SQLErrorCode.INVALID_PARAMETER_VALUE, SQLErrorCode.INVALID_REGULAR_EXPRESSION, SQLErrorCode.INVALID_ROW_COUNT_IN_LIMIT_CLAUSE, SQLErrorCode.INVALID_ROW_COUNT_IN_RESULT_OFFSET_CLAUSE, SQLErrorCode.INVALID_TIME_ZONE_DISPLACEMENT_VALUE, SQLErrorCode.INVALID_USE_OF_ESCAPE_CHARACTER, SQLErrorCode.MOST_SPECIFIC_TYPE_MISMATCH, SQLErrorCode.NULL_VALUE_NOT_ALLOWED, SQLErrorCode.NULL_VALUE_NO_INDICATOR_PARAMETER, SQLErrorCode.NUMERIC_VALUE_OUT_OF_RANGE, SQLErrorCode.STRING_DATA_LENGTH_MISMATCH, SQLErrorCode.SUBSTRING_ERROR, SQLErrorCode.TRIM_ERROR, SQLErrorCode.UNTERMINATED_C_STRING, SQLErrorCode.ZERO_LENGTH_CHARACTER_STRING, SQLErrorCode.FLOATING_POINT_EXCEPTION, SQLErrorCode.INVALID_TEXT_REPRESENTATION, SQLErrorCode.INVALID_BINARY_REPRESENTATION, SQLErrorCode.BAD_COPY_FILE_FORMAT, SQLErrorCode.UNTRANSLATABLE_CHARACTER, SQLErrorCode.NOT_AN_XML_DOCUMENT, SQLErrorCode.INVALID_XML_DOCUMENT, SQLErrorCode.INVALID_XML_CONTENT, SQLErrorCode.INVALID_XML_COMMENT, SQLErrorCode.INVALID_XML_PROCESSING_INSTRUCTION, SQLErrorCode.INTEGRITY_CONSTRAINT_VIOLATION, SQLErrorCode.RESTRICT_VIOLATION, SQLErrorCode.NOT_NULL_VIOLATION, SQLErrorCode.FOREIGN_KEY_VIOLATION, SQLErrorCode.UNIQUE_VIOLATION, SQLErrorCode.CHECK_VIOLATION, SQLErrorCode.EXCLUSION_VIOLATION, SQLErrorCode.INVALID_CURSOR_STATE, SQLErrorCode.INVALID_TRANSACTION_STATE, SQLErrorCode.ACTIVE_SQL_TRANSACTION, SQLErrorCode.BRANCH_TRANSACTION_ALREADY_ACTIVE, SQLErrorCode.HELD_CURSOR_REQUIRES_SAME_ISOLATION_LEVEL, SQLErrorCode.INAPPROPRIATE_ACCESS_MODE_FOR_BRANCH_TRANSACTION, SQLErrorCode.INAPPROPRIATE_ISOLATION_LEVEL_FOR_BRANCH_TRANSACTION, SQLErrorCode.NO_ACTIVE_SQL_TRANSACTION_FOR_BRANCH_TRANSACTION, SQLErrorCode.READ_ONLY_SQL_TRANSACTION, SQLErrorCode.SCHEMA_AND_DATA_STATEMENT_MIXING_NOT_SUPPORTED, SQLErrorCode.NO_ACTIVE_SQL_TRANSACTION, SQLErrorCode.IN_FAILED_SQL_TRANSACTION, SQLErrorCode.INVALID_SQL_STATEMENT_NAME, SQLErrorCode.TRIGGERED_DATA_CHANGE_VIOLATION, SQLErrorCode.INVALID_AUTHORIZATION_SPECIFICATION, SQLErrorCode.INVALID_PASSWORD, SQLErrorCode.DEPENDENT_PRIVILEGE_DESCRIPTORS_STILL_EXIST, SQLErrorCode.DEPENDENT_OBJECTS_STILL_EXIST, SQLErrorCode.INVALID_TRANSACTION_TERMINATION, SQLErrorCode.SQL_ROUTINE_EXCEPTION, SQLErrorCode.FUNCTION_EXECUTED_NO_RETURN_STATEMENT, SQLErrorCode.MODIFYING_SQL_DATA_NOT_PERMITTED, SQLErrorCode.PROHIBITED_SQL_STATEMENT_ATTEMPTED, SQLErrorCode.READING_SQL_DATA_NOT_PERMITTED, SQLErrorCode.INVALID_CURSOR_NAME, SQLErrorCode.EXTERNAL_ROUTINE_EXCEPTION, SQLErrorCode.CONTAINING_SQL_NOT_PERMITTED, SQLErrorCode.EXTERNAL_ROUTINE_INVOCATION_EXCEPTION, SQLErrorCode.INVALID_SQLSTATE_RETURNED, SQLErrorCode.TRIGGER_PROTOCOL_VIOLATED, SQLErrorCode.SRF_PROTOCOL_VIOLATED, SQLErrorCode.SAVEPOINT_EXCEPTION, SQLErrorCode.INVALID_SAVEPOINT_SPECIFICATION, SQLErrorCode.INVALID_CATALOG_NAME, SQLErrorCode.INVALID_SCHEMA_NAME, SQLErrorCode.TRANSACTION_ROLLBACK, SQLErrorCode.TRANSACTION_INTEGRITY_CONSTRAINT_VIOLATION, SQLErrorCode.SERIALIZATION_FAILURE, SQLErrorCode.STATEMENT_COMPLETION_UNKNOWN, SQLErrorCode.DEADLOCK_DETECTED, SQLErrorCode.SYNTAX_ERROR_OR_ACCESS_RULE_VIOLATION, SQLErrorCode.SYNTAX_ERROR, SQLErrorCode.INSUFFICIENT_PRIVILEGE, SQLErrorCode.CANNOT_COERCE, SQLErrorCode.GROUPING_ERROR, SQLErrorCode.WINDOWING_ERROR, SQLErrorCode.INVALID_RECURSION, SQLErrorCode.INVALID_FOREIGN_KEY, SQLErrorCode.INVALID_NAME, SQLErrorCode.NAME_TOO_LONG, SQLErrorCode.RESERVED_NAME, SQLErrorCode.DATATYPE_MISMATCH, SQLErrorCode.INDETERMINATE_DATATYPE, SQLErrorCode.COLLATION_MISMATCH, SQLErrorCode.INDETERMINATE_COLLATION, SQLErrorCode.WRONG_OBJECT_TYPE, SQLErrorCode.UNDEFINED_COLUMN, SQLErrorCode.UNDEFINED_FUNCTION, SQLErrorCode.UNDEFINED_TABLE, SQLErrorCode.UNDEFINED_PARAMETER, SQLErrorCode.UNDEFINED_OBJECT, SQLErrorCode.DUPLICATE_COLUMN, SQLErrorCode.DUPLICATE_CURSOR, SQLErrorCode.DUPLICATE_DATABASE, SQLErrorCode.DUPLICATE_FUNCTION, SQLErrorCode.DUPLICATE_PREPARED_STATEMENT, SQLErrorCode.DUPLICATE_SCHEMA, SQLErrorCode.DUPLICATE_TABLE, SQLErrorCode.DUPLICATE_ALIAS, SQLErrorCode.DUPLICATE_OBJECT, SQLErrorCode.AMBIGUOUS_COLUMN, SQLErrorCode.AMBIGUOUS_FUNCTION, SQLErrorCode.AMBIGUOUS_PARAMETER, SQLErrorCode.AMBIGUOUS_ALIAS, SQLErrorCode.INVALID_COLUMN_REFERENCE, SQLErrorCode.INVALID_COLUMN_DEFINITION, SQLErrorCode.INVALID_CURSOR_DEFINITION, SQLErrorCode.INVALID_DATABASE_DEFINITION, SQLErrorCode.INVALID_FUNCTION_DEFINITION, SQLErrorCode.INVALID_PREPARED_STATEMENT_DEFINITION, SQLErrorCode.INVALID_SCHEMA_DEFINITION, SQLErrorCode.INVALID_TABLE_DEFINITION, SQLErrorCode.INVALID_OBJECT_DEFINITION, SQLErrorCode.WITH_CHECK_OPTION_VIOLATION, SQLErrorCode.INSUFFICIENT_RESOURCES, SQLErrorCode.DISK_FULL, SQLErrorCode.OUT_OF_MEMORY, SQLErrorCode.TOO_MANY_CONNECTIONS, SQLErrorCode.CONFIGURATION_LIMIT_EXCEEDED, SQLErrorCode.PROGRAM_LIMIT_EXCEEDED, SQLErrorCode.STATEMENT_TOO_COMPLEX, SQLErrorCode.TOO_MANY_COLUMNS, SQLErrorCode.TOO_MANY_ARGUMENTS, SQLErrorCode.OBJECT_NOT_IN_PREREQUISITE_STATE, SQLErrorCode.OBJECT_IN_USE, SQLErrorCode.CANT_CHANGE_RUNTIME_PARAM, SQLErrorCode.LOCK_NOT_AVAILABLE, SQLErrorCode.OPERATOR_INTERVENTION, SQLErrorCode.QUERY_CANCELED, SQLErrorCode.ADMIN_SHUTDOWN, SQLErrorCode.CRASH_SHUTDOWN, SQLErrorCode.CANNOT_CONNECT_NOW, SQLErrorCode.DATABASE_DROPPED, SQLErrorCode.SYSTEM_ERROR, SQLErrorCode.IO_ERROR, SQLErrorCode.UNDEFINED_FILE, SQLErrorCode.DUPLICATE_FILE, SQLErrorCode.CONFIG_FILE_ERROR, SQLErrorCode.LOCK_FILE_EXISTS, SQLErrorCode.FDW_ERROR, SQLErrorCode.FDW_COLUMN_NAME_NOT_FOUND, SQLErrorCode.FDW_DYNAMIC_PARAMETER_VALUE_NEEDED, SQLErrorCode.FDW_FUNCTION_SEQUENCE_ERROR, SQLErrorCode.FDW_INCONSISTENT_DESCRIPTOR_INFORMATION, SQLErrorCode.FDW_INVALID_ATTRIBUTE_VALUE, SQLErrorCode.FDW_INVALID_COLUMN_NAME, SQLErrorCode.FDW_INVALID_COLUMN_NUMBER, SQLErrorCode.FDW_INVALID_DATA_TYPE, SQLErrorCode.FDW_INVALID_DATA_TYPE_DESCRIPTORS, SQLErrorCode.FDW_INVALID_DESCRIPTOR_FIELD_IDENTIFIER, SQLErrorCode.FDW_INVALID_HANDLE, SQLErrorCode.FDW_INVALID_OPTION_INDEX, SQLErrorCode.FDW_INVALID_OPTION_NAME, SQLErrorCode.FDW_INVALID_STRING_LENGTH_OR_BUFFER_LENGTH, SQLErrorCode.FDW_INVALID_STRING_FORMAT, SQLErrorCode.FDW_INVALID_USE_OF_NULL_POINTER, SQLErrorCode.FDW_TOO_MANY_HANDLES, SQLErrorCode.FDW_OUT_OF_MEMORY, SQLErrorCode.FDW_NO_SCHEMAS, SQLErrorCode.FDW_OPTION_NAME_NOT_FOUND, SQLErrorCode.FDW_REPLY_HANDLE, SQLErrorCode.FDW_SCHEMA_NOT_FOUND, SQLErrorCode.FDW_TABLE_NOT_FOUND, SQLErrorCode.FDW_UNABLE_TO_CREATE_EXECUTION, SQLErrorCode.FDW_UNABLE_TO_CREATE_REPLY, SQLErrorCode.FDW_UNABLE_TO_ESTABLISH_CONNECTION, SQLErrorCode.PLPGSQL_ERROR, SQLErrorCode.RAISE_EXCEPTION, SQLErrorCode.NO_DATA_FOUND, SQLErrorCode.TOO_MANY_ROWS, SQLErrorCode.INTERNAL_ERROR, SQLErrorCode.DATA_CORRUPTED, SQLErrorCode.INDEX_CORRUPTED]);
